packages/cadl-ranch-specs/http/models/inheritance/main.cadl

```diff
@@ -1,60 +1,112 @@
-@doc("Example base class that has a discriminator property.")
-model BaseClassWithDiscriminator extends BaseClass {}
-// Note: discriminator property is not defined on parent
+@doc("Base model for three levels inheritance with discriminator.")
+@discriminator("kind")
+model Fish {
+  age?: int32;
+}
+
+@doc("The second level model in three levels inheritance with discriminator and defined a new discriminator.")
+@discriminator("sharktype")
+model Shark extends Fish {
+  kind: "shark";
+}
 
-model DerivedFromBaseClassWithDiscriminatorA extends BaseClassWithDiscriminator {
-  discriminatorProperty: "A";
+@doc("The second level model in three levels inheritance with discriminator.")
+model Salmon extends Fish {
+  kind: "salmon";
+  friends?: Fish[];
+  hate?: Record<Fish>;
+  partner?: Fish;
 }
 
-model DerivedFromBaseClassWithDiscriminatorB extends BaseClassWithDiscriminator {
-  discriminatorProperty: "B";
+@doc("The third level model A in the three levels inheritance with discriminators.")
+model SawShark extends Shark {
+  sharktype: "saw";
 }
 
-@doc("Illustrates case where a basic model has polymorphic properties.")
-model ModelWithPolymorphicProperty {
-  @doc("Example polymorphic type property.")
-  polymorphicProperty: BaseClassWithDiscriminator;
+@doc("The third level model B in the non-discriminator three levels inheritance with discriminators.")
+model GoblinShark extends Shark {
+  sharktype: "goblin";
 }
 
 @scenario
-@scenarioDoc("Generate, send, and receive round-trip inherited model.")
-@route("/model")
+@scenarioDoc("Generate and send model. The valid input value is {name: \"abc\", age: 32, smart: true}")
+@route("/valid")
+@post
+op sendInheritanceValid(@body input: Siamese): OkResponse | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate and receive model. The return value is {name: \"abc\", age: 32, smart: true}")
+@route("/valid")
+@get
+op getInheritanceValid(): Siamese | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate, send, and receive round-trip bottom model.")
+@route("/valid")
 @put
-op setValue(@body input: BaseClass): BaseClass | ErrorResponse;
+op setInheritanceValid(@body input: Siamese): Siamese | ErrorResponse;
 
 @scenario
-@scenarioDoc("Generate, send, and receive round-trip model with a polymorphic property.")
-@route("/property")
+@route("/invalid")
+@scenarioDoc("Get a basic model that is invalid for the local strong kind.")
+@get
+op getInheritanceInvalid(): Pet | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate, send, and receive round-trip model that optional property is not presented.")
+@route("/empty")
 @put
-op setValueWithPolymorphicProperty(
-  @body input: ModelWithPolymorphicProperty
-): ModelWithPolymorphicProperty | ErrorResponse;
+op setInheritanceEmpty(@body input: Siamese): Pet | ErrorResponse;
+
+@scenario
+@scenarioDoc("Get a basic model that optional property is sent as null.")
+@route("/null")
+@get
+op getInheritanceWithNullProperty(): Pet | ErrorResponse;
```

Comment:
what are we trying to test here? Sorry I'm just trying to figure out what scenario we're trying to test and if it's a good scenario to test

Reply:
It's for testing what happen when service return an optional property as null.
Moved this test case to `property-optional``in this PR https://github.com/Azure/cadl-ranch/pull/92